File: proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/masterdata/BusinessAreaMasterDataFactory.java
package com.jasonzqshen.familyaccounting.core.masterdata;

import org.w3c.dom.Element;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.exception.IdentityInvalidChar;
import com.jasonzqshen.familyaccounting.core.exception.IdentityNoData;
import com.jasonzqshen.familyaccounting.core.exception.IdentityTooLong;
import com.jasonzqshen.familyaccounting.core.exception.MasterDataIdentityExists;
import com.jasonzqshen.familyaccounting.core.exception.NullValueNotAcceptable;
import com.jasonzqshen.familyaccounting.core.exception.ParametersException;
import com.jasonzqshen.familyaccounting.core.exception.format.MasterDataFileFormatException;
import com.jasonzqshen.familyaccounting.core.exception.runtime.SystemException;
import com.jasonzqshen.familyaccounting.core.utils.CoreMessage;
import com.jasonzqshen.familyaccounting.core.utils.CriticalLevel;
import com.jasonzqshen.familyaccounting.core.utils.MessageType;
import com.jasonzqshen.familyaccounting.core.utils.StringUtility;

public class BusinessAreaMasterDataFactory extends MasterDataFactoryBase {

	public BusinessAreaMasterDataFactory(CoreDriver coreDriver,
			MasterDataManagement management) {
		super(coreDriver, management);
	}

	@Override
	public MasterDataBase createNewMasterDataBase(MasterDataIdentity identity,
			String descp, Object... objects) throws SystemException,
			ParametersException, MasterDataIdentityExists {
		if (objects.length != 1) {
			throw new ParametersException(String.format(
					CoreMessage.ERR_PARAMETER_LENGTH, 1, objects.length));
		}

		// check identity is duplicated
		if (_list.containsKey(identity)) {
			throw new MasterDataIdentityExists();
		}

		CriticalLevel l = null;
		if (!(objects[0] instanceof CriticalLevel)) {
			throw new ParametersException(String.format(
					CoreMessage.ERR_PARAMETER_TYPE,
					CriticalLevel.class.getName()));
		}
		l = (CriticalLevel) objects[0];
		BusinessAreaMasterData businessArea;
		try {
			businessArea = new BusinessAreaMasterData(_coreDriver, _management, identity,
					descp, l);
		} catch (NullValueNotAcceptable e) {
			throw new SystemException(e);
		}

		this._containDirtyData = true;
		this._list.put(identity, businessArea);

		// raise create master data
		_coreDriver.getListenersManagement().createMasterData(this,
				businessArea);
		_coreDriver.logDebugInfo(
				this.getClass(),
				59,
				String.format("Create business area (%s).",
						businessArea.toXML()), MessageType.INFO);
		return businessArea;
	}

	@Override
	public MasterDataBase parseMasterData(CoreDriver coreDriver, Element elem)
			throws MasterDataFileFormatException {
		String id = elem.getAttribute(MasterDataUtils.XML_ID);
		String descp = elem.getAttribute(MasterDataUtils.XML_DESCP);
		String criticalLevel = elem
				.getAttribute(MasterDataUtils.XML_CRITICAL_LEVEL);
		// check attribute
		if (StringUtility.isNullOrEmpty(criticalLevel)) {
			throw new MasterDataFileFormatException(
					MasterDataType.BUSINESS_AREA);
		}

		CriticalLevel l = CriticalLevel.parse(criticalLevel.charAt(0));
		try {
			MasterDataIdentity identity = new MasterDataIdentity(
					id.toCharArray());

			BusinessAreaMasterData businessArea = (BusinessAreaMasterData) this
					.createNewMasterDataBase(identity, descp, l);

			_coreDriver.logDebugInfo(
					this.getClass(),
					130,
					String.format("Parse business area (%s).",
							businessArea.toXML()), MessageType.INFO);
			return businessArea;
		} catch (IdentityTooLong e) {
			_coreDriver.logDebugInfo(this.getClass(), 150,
					"Master data identity is too long.", MessageType.ERROR);
			throw new MasterDataFileFormatException(
					MasterDataType.BUSINESS_AREA);
		} catch (IdentityNoData e) {
			_coreDriver
					.logDebugInfo(this.getClass(), 154,
							"Master data identity is with no value.",
							MessageType.ERROR);
			throw new MasterDataFileFormatException(
					MasterDataType.BUSINESS_AREA);
		} catch (IdentityInvalidChar e) {
			_coreDriver.logDebugInfo(this.getClass(), 160,
					"Invalid character in identity.", MessageType.ERROR);
			throw new MasterDataFileFormatException(
					MasterDataType.BUSINESS_AREA);
		} catch (ParametersException e) {
			_coreDriver.logDebugInfo(this.getClass(), 164,
					"Function parameter set error: " + e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (MasterDataIdentityExists e) {
			_coreDriver.logDebugInfo(this.getClass(), 168,
					"Master data identity duplicated.", MessageType.ERROR);
			throw new MasterDataFileFormatException(
					MasterDataType.BUSINESS_AREA);
		}

	}

}

File: branches/V1.0/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/MasterData/BusinessAreaMasterDataFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.MasterData
{
    public class BusinessAreaMasterDataFactory : MasterDataFactoryBase
    {
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="management"></param>
        public BusinessAreaMasterDataFactory(CoreDriver coreDriver,
                MasterDataManagement management)
            : base(coreDriver, management)
        {
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="management"></param>
        /// <param name="id"></param>
        /// <param name="descp"></param>
        /// <param name="level"></param>
        /// <exception cref="ArgumentNullException">argument is null</exception>
        public override MasterDataBase CreateNewMasterDataBase(MasterDataIdentity identity,
                String descp, params Object[] objects)
        {
            if (objects.Length != 1)
            {
                throw new ParametersException(1, objects.Length);
            }

            // check identity is duplicated
            if (_list.ContainsKey(identity))
            {
                throw new MasterDataIdentityExists();
            }

            if (!(objects[0] is CriticalLevel))
            {
                throw new ParametersException();
            }
            CriticalLevel l = (CriticalLevel)objects[0];

            BusinessAreaMasterData businessArea;
            try
            {
                businessArea = new BusinessAreaMasterData(_coreDriver, _management, identity,
                        descp, l);
            }
            catch (ArgumentNullException e)
            {
                throw new SystemException(e);
            }

            this._containDirtyData = true;
            this._list.Add(identity, businessArea);

            // raise create master data
            _coreDriver.ListenerMgmt.CreateMasterData(this,
                    businessArea);
            _coreDriver.logDebugInfo(
                    this.GetType(),
                    59,
                    String.Format("Create business area (%s).",
                            businessArea.Identity.ToString()), MessageType.INFO);
            return businessArea;
        }

        /// <summary>
        /// Parse master data from XML
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="elem"></param>
        /// <returns></returns>
        /// <exception cref="MasterDataFileFormatException">Master Data file exception</exception>
        /// <exception cref="ArgumentNullException">Argument is null</exception>
        /// <exception cref="SystemException">Bug</exception>
        public override MasterDataBase ParseMasterData(CoreDriver coreDriver, XElement elem)
        {
            XAttribute id = elem.Attribute(MasterDataUtils.XML_ID);
            XAttribute descp = elem.Attribute(MasterDataUtils.XML_DESCP);
            XAttribute criticalLevel = elem
                    .Attribute(MasterDataUtils.XML_CRITICAL_LEVEL);
            // check attribute
            if (criticalLevel == null)
            {
                throw new MasterDataFileFormatException(
                        MasterDataType.BUSINESS_AREA);
            }
            if (criticalLevel.Value.Length != 1
                || (criticalLevel.Value[0] != (char)CriticalLevel.HIGH
                && criticalLevel.Value[0] != (char)CriticalLevel.MEDIUM
                && criticalLevel.Value[0] != (char)CriticalLevel.LOW))
            {
                _coreDriver
                        .logDebugInfo(this.GetType(), 154,
                                "Format of critical level is error",
                                MessageType.ERRO);
                throw new MasterDataFileFormatException(
                        MasterDataType.BUSINESS_AREA);
            }
            CriticalLevel l = (CriticalLevel)criticalLevel.Value[0];

            MasterDataIdentity identity;
            try
            {
                identity = new MasterDataIdentity(id.Value);
            }
            catch (Exception e)
            {
                _coreDriver
                        .logDebugInfo(this.GetType(), 154,
                                e.Message,
                                MessageType.ERRO);
                throw new MasterDataFileFormatException(
                        MasterDataType.BUSINESS_AREA);
            }

            try
            {
                BusinessAreaMasterData businessArea = (BusinessAreaMasterData)this
                            .CreateNewMasterDataBase(identity, descp.Value, l);

                _coreDriver.logDebugInfo(
                        this.GetType(),
                        130,
                        String.Format("Parse business area ({0}).",
                                businessArea.Identity.ToString()), MessageType.INFO);
                return businessArea;
            }
            catch (Exception e)
            {
                throw new SystemException(e);
            }

        }

    }
}
